File: android/Gutenberg/src/test/java/org/wordpress/gutenberg/http/RequestBodyTests.kt
package org.wordpress.gutenberg.http

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

class RequestBodyTests {

    // MARK: - InMemory

    @Test
    fun `InMemory inputStream returns correct data`() {
        val data = "hello world".toByteArray()
        val body = RequestBody.InMemory(data)

        val result = body.inputStream().use { it.readBytes() }
        assertArrayEquals(data, result)
    }

    @Test
    fun `InMemory readBytes returns copy of data`() {
        val data = "hello".toByteArray()
        val body = RequestBody.InMemory(data)
        val copy = body.readBytes()

        assertArrayEquals(data, copy)
        // Mutating the copy should not affect the original.
        copy[0] = 'X'.code.toByte()
        assertEquals('h'.code.toByte(), data[0])
    }

    @Test
    fun `InMemory size returns correct value`() {
        val body = RequestBody.InMemory("hello".toByteArray())
        assertEquals(5L, body.size)
    }

    @Test
    fun `InMemory inMemoryData returns non-null`() {
        val body = RequestBody.InMemory("test".toByteArray())
        assertNotNull(body.inMemoryData)
        assertEquals(4, body.inMemoryData!!.size)
    }

    @Test
    fun `InMemory file returns null`() {
        val body = RequestBody.InMemory("test".toByteArray())
        assertNull(body.file)
    }

    @Test
    fun `InMemory equality with same data`() {
        val a = RequestBody.InMemory("hello".toByteArray())
        val b = RequestBody.InMemory("hello".toByteArray())
        assertEquals(a, b)
        assertEquals(a.hashCode(), b.hashCode())
    }

    @Test
    fun `InMemory inequality with different data`() {
        val a = RequestBody.InMemory("hello".toByteArray())
        val b = RequestBody.InMemory("world".toByteArray())
        assertNotEquals(a, b)
    }

    @Test
    fun `InMemory empty body`() {
        val body = RequestBody.InMemory(ByteArray(0))
        assertEquals(0L, body.size)
        assertArrayEquals(ByteArray(0), body.readBytes())
        assertArrayEquals(ByteArray(0), body.inputStream().use { it.readBytes() })
    }

    // MARK: - FileBacked

    @Test
    fun `FileBacked inputStream returns correct data`() {
        val data = "file-backed content".toByteArray()
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes(data)

        val body = RequestBody.FileBacked(file = file, fileOffset = 0, size = data.size.toLong())
        val result = body.inputStream().use { it.readBytes() }
        assertArrayEquals(data, result)
    }

    @Test
    fun `FileBacked readBytes returns correct data`() {
        val data = "readBytes test".toByteArray()
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes(data)

        val body = RequestBody.FileBacked(file = file, fileOffset = 0, size = data.size.toLong())
        assertArrayEquals(data, body.readBytes())
    }

    @Test
    fun `FileBacked with offset reads correct slice`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("GARBAGE_hello world_TRAILING".toByteArray())

        // Read "hello world" starting at offset 8, length 11.
        val body = RequestBody.FileBacked(file = file, fileOffset = 8, size = 11)

        assertArrayEquals("hello world".toByteArray(), body.readBytes())
        assertArrayEquals("hello world".toByteArray(), body.inputStream().use { it.readBytes() })
    }

    @Test
    fun `FileBacked inputStream respects offset and size boundary`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("AAAbbbCCC".toByteArray())

        val body = RequestBody.FileBacked(file = file, fileOffset = 3, size = 3)

        val stream = body.inputStream()
        val result = stream.use { it.readBytes() }
        assertArrayEquals("bbb".toByteArray(), result)
    }

    @Test
    fun `FileBacked single byte read`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("ABCDE".toByteArray())

        val body = RequestBody.FileBacked(file = file, fileOffset = 2, size = 1)
        val stream = body.inputStream()
        assertEquals('C'.code, stream.read())
        assertEquals(-1, stream.read()) // Past the size boundary.
        stream.close()
    }

    @Test
    fun `FileBacked binary data preserved`() {
        val binaryData = ByteArray(512) { it.toByte() }
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes(binaryData)

        val body = RequestBody.FileBacked(file = file, fileOffset = 0, size = binaryData.size.toLong())
        assertArrayEquals(binaryData, body.readBytes())
        assertArrayEquals(binaryData, body.inputStream().use { it.readBytes() })
    }

    @Test
    fun `FileBacked zero-length body`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("content".toByteArray())

        val body = RequestBody.FileBacked(file = file, fileOffset = 3, size = 0)
        assertEquals(0L, body.size)
        assertArrayEquals(ByteArray(0), body.readBytes())
    }

    @Test
    fun `FileBacked multiple streams read independently`() {
        val data = "independent reads".toByteArray()
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes(data)

        val body = RequestBody.FileBacked(file = file, fileOffset = 0, size = data.size.toLong())

        val stream1 = body.inputStream()
        val stream2 = body.inputStream()

        val result1 = stream1.use { it.readBytes() }
        val result2 = stream2.use { it.readBytes() }

        assertArrayEquals(data, result1)
        assertArrayEquals(data, result2)
    }

    @Test
    fun `FileBacked inMemoryData returns null`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("test".toByteArray())

        val body = RequestBody.FileBacked(file = file, fileOffset = 0, size = 4)
        assertNull(body.inMemoryData)
    }

    @Test
    fun `FileBacked equality with same file and range`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("content".toByteArray())

        val a = RequestBody.FileBacked(file = file, fileOffset = 1, size = 3)
        val b = RequestBody.FileBacked(file = file, fileOffset = 1, size = 3)
        assertEquals(a, b)
        assertEquals(a.hashCode(), b.hashCode())
    }

    @Test
    fun `FileBacked inequality with different offset`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("content".toByteArray())

        val a = RequestBody.FileBacked(file = file, fileOffset = 0, size = 3)
        val b = RequestBody.FileBacked(file = file, fileOffset = 1, size = 3)
        assertNotEquals(a, b)
    }

    @Test
    fun `FileBacked inequality with different size`() {
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes("content".toByteArray())

        val a = RequestBody.FileBacked(file = file, fileOffset = 0, size = 3)
        val b = RequestBody.FileBacked(file = file, fileOffset = 0, size = 5)
        assertNotEquals(a, b)
    }

    // MARK: - Cross-type inequality

    @Test
    fun `InMemory and FileBacked are not equal even with same content`() {
        val data = "same content".toByteArray()
        val file = File.createTempFile("rb-test-", null)
        file.deleteOnExit()
        file.writeBytes(data)

        val inMemory = RequestBody.InMemory(data)
        val fileBacked = RequestBody.FileBacked(file = file, fileOffset = 0, size = data.size.toLong())
        assertNotEquals(inMemory as RequestBody, fileBacked as RequestBody)
    }
}

File: ios/Tests/GutenbergKitHTTPTests/RequestBodyTests.swift
import Foundation
import Testing
@testable import GutenbergKitHTTP

@Suite("RequestBody")
struct RequestBodyTests {

    // MARK: - makeInputStream (Data-backed)

    @Test("makeInputStream returns working stream for data-backed body")
    func dataBackedStream() throws {
        let expected = Data("hello world".utf8)
        let body = RequestBody(data: expected)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == expected)
    }

    // MARK: - makeInputStream (File-backed)

    @Test("makeInputStream returns working stream for file-backed body")
    func fileBackedStream() throws {
        let expected = Data("file contents".utf8)
        let url = makeTemporaryFile(contents: expected)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == expected)
    }

    @Test("makeInputStream throws fileNoSuchFile for missing file")
    func throwsForMissingFile() {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("nonexistent-\(UUID().uuidString)")
        let body = RequestBody(fileURL: url)

        #expect(throws: CocoaError.self) {
            let error = try body.makeInputStream()
            _ = error  // suppress unused warning
        }
    }

    @Test("makeInputStream throws fileReadInvalidFileName for directory")
    func throwsForDirectory() throws {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("dir-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url)

        #expect(throws: CocoaError.self) {
            let error = try body.makeInputStream()
            _ = error
        }
    }

    // MARK: - makeInputStream (FileSlice-backed)

    @Test("fileSlice stream reads correct byte range")
    func fileSliceReadsCorrectRange() throws {
        let contents = Data("Hello, World!".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        // Slice "World" (offset 7, length 5)
        let body = RequestBody(fileURL: url, offset: 7, length: 5)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == Data("World".utf8))
    }

    @Test("fileSlice stream reads from beginning when offset is zero")
    func fileSliceFromBeginning() throws {
        let contents = Data("ABCDEFGHIJ".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 0, length: 3)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == Data("ABC".utf8))
    }

    @Test("fileSlice stream reads to end of file")
    func fileSliceToEnd() throws {
        let contents = Data("ABCDEFGHIJ".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 7, length: 3)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == Data("HIJ".utf8))
    }

    @Test("fileSlice stream reads entire file when slice covers all bytes")
    func fileSliceEntireFile() throws {
        let contents = Data("complete file".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 0, length: contents.count)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == contents)
    }

    @Test("fileSlice stream reads single byte")
    func fileSliceSingleByte() throws {
        let contents = Data("ABCDE".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 2, length: 1)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == Data("C".utf8))
    }

    @Test("fileSlice stream works with reads smaller than slice length")
    func fileSliceSmallReads() throws {
        // Create a body larger than a typical small read buffer
        let contents = Data(repeating: 0x42, count: 4096)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 100, length: 3000)

        let stream = try body.makeInputStream()
        // Read in small chunks to exercise the bounded read logic
        let result = readAllWithBufferSize(stream, bufferSize: 128)
        #expect(result == contents[100..<3100])
    }

    @Test("fileSlice stream returns 0 after exhaustion")
    func fileSliceReadAfterExhaustion() throws {
        let contents = Data("AB".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 0, length: 2)
        let stream = try body.makeInputStream()
        stream.open()

        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: 10)
        defer { buffer.deallocate() }

        // First read gets all bytes
        let n1 = stream.read(buffer, maxLength: 10)
        #expect(n1 == 2)

        // Second read returns 0 (at end)
        let n2 = stream.read(buffer, maxLength: 10)
        #expect(n2 == 0)

        stream.close()
    }

    @Test("fileSlice init throws for missing file")
    func fileSliceThrowsForMissingFile() {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("nonexistent-\(UUID().uuidString)")
        let body = RequestBody(fileURL: url, offset: 0, length: 10)

        #expect(throws: Error.self) {
            _ = try body.makeInputStream()
        }
    }

    @Test("fileSlice stream with zero length returns no data")
    func fileSliceZeroLength() throws {
        let contents = Data("ABCDE".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 2, length: 0)

        let stream = try body.makeInputStream()
        #expect(readAll(stream) == Data())
    }

    @Test("fileSlice stream does not read beyond slice boundary")
    func fileSliceDoesNotReadBeyondBoundary() throws {
        let contents = Data("ABCDEFGHIJ".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        // Slice is "CDE" (offset 2, length 3) — must not return "FGHIJ"
        let body = RequestBody(fileURL: url, offset: 2, length: 3)

        let stream = try body.makeInputStream()
        let result = readAll(stream)
        #expect(result == Data("CDE".utf8))
        #expect(result.count == 3)
    }

    @Test("fileSlice stream with binary data preserves all bytes")
    func fileSliceBinaryData() throws {
        // All byte values 0x00-0xFF
        let contents = Data(0...255)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body = RequestBody(fileURL: url, offset: 100, length: 50)

        let stream = try body.makeInputStream()
        let result = readAll(stream)
        #expect(result == Data(100..<150))
    }

    @Test("multiple fileSlice streams from same file read independently")
    func fileSliceMultipleStreamsIndependent() throws {
        let contents = Data("ABCDEFGHIJKLMNOP".utf8)
        let url = makeTemporaryFile(contents: contents)
        defer { try? FileManager.default.removeItem(at: url) }

        let body1 = RequestBody(fileURL: url, offset: 0, length: 4)
        let body2 = RequestBody(fileURL: url, offset: 8, length: 4)

        let stream1 = try body1.makeInputStream()
        let stream2 = try body2.makeInputStream()

        #expect(readAll(stream1) == Data("ABCD".utf8))
        #expect(readAll(stream2) == Data("IJKL".utf8))
    }

    // MARK: - Equatable

    @Test("data-backed bodies with same data are equal")
    func dataEquality() {
        let data = Data("same".utf8)
        #expect(RequestBody(data: data) == RequestBody(data: data))
    }

    @Test("data-backed bodies with different data are not equal")
    func dataInequality() {
        #expect(RequestBody(data: Data("a".utf8)) != RequestBody(data: Data("b".utf8)))
    }

    @Test("file-backed bodies with same URL are equal")
    func fileEquality() {
        let url = URL(fileURLWithPath: "/tmp/same-file")
        #expect(RequestBody(fileURL: url) == RequestBody(fileURL: url))
    }

    @Test("data-backed and file-backed bodies are not equal")
    func dataVsFileInequality() {
        let data = Data("hello".utf8)
        let url = URL(fileURLWithPath: "/tmp/hello")
        #expect(RequestBody(data: data) != RequestBody(fileURL: url))
    }

    // MARK: - URLSession integration (file-slice via httpBodyStream)

    #if canImport(Network)
    @Test("fileSlice body sent via URLSession httpBodyStream delivers correct bytes")
    func fileSliceBodyStreamWorksWithURLSession() async throws {
        // Write a known payload to a temp file
        let payload = Data("The quick brown fox jumps over the lazy dog".utf8)
        let fileURL = makeTemporaryFile(contents: payload)
        defer { try? FileManager.default.removeItem(at: fileURL) }

        // Slice out "brown fox" (offset 10, length 9)
        let expectedSlice = Data("brown fox".utf8)
        let body = RequestBody(fileURL: fileURL, offset: 10, length: 9)

        // Start an echo server that returns the request body as the response
        let server = try await HTTPServer.start(
            name: "body-echo",
            requiresAuthentication: false
        ) { request in
            guard let body = request.parsed.body,
                  let data = try? await body.data else {
                return HTTPResponse(status: 200, body: Data())
            }
            return HTTPResponse(status: 200, body: data)
        }
        defer { server.stop() }

        // Build a URLRequest using the same code path as the proxy:
        // this assigns body.makeInputStream() to request.httpBodyStream.
        let baseURL = URL(string: "http://127.0.0.1:\(server.port)")!
        let parsed = ParsedHTTPRequest.complete(
            method: "POST",
            target: "/echo",
            httpVersion: "HTTP/1.1",
            headers: ["Host": "localhost", "Content-Length": "9"],
            body: body
        )
        var request = try #require(parsed.urlRequest(relativeTo: baseURL))
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")

        let (responseData, response) = try await URLSession.shared.data(for: request)
        let http = try #require(response as? HTTPURLResponse)

        #expect(http.statusCode == 200)
        // This assertion would fail with the old FileSliceInputStream subclass
        // because URLSession reads from the empty Data() superclass instead of
        // the overridden read(_:maxLength:) — resulting in an empty body.
        #expect(responseData == expectedSlice)
    }
    #endif

    // MARK: - Helpers

    private func makeTemporaryFile(contents: Data) -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("RequestBodyTests-\(UUID().uuidString)")
        FileManager.default.createFile(atPath: url.path, contents: contents)
        return url
    }

    private func readAll(_ stream: InputStream) -> Data {
        readAllWithBufferSize(stream, bufferSize: 1024)
    }

    private func readAllWithBufferSize(_ stream: InputStream, bufferSize: Int) -> Data {
        stream.open()
        defer { stream.close() }

        var data = Data()
        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
        defer { buffer.deallocate() }

        // Use read() directly instead of hasBytesAvailable to avoid race
        // conditions with bound stream pairs, where data from the writer
        // thread may not have arrived yet when hasBytesAvailable is checked.
        while true {
            let bytesRead = stream.read(buffer, maxLength: bufferSize)
            if bytesRead <= 0 { break }
            data.append(buffer, count: bytesRead)
        }

        return data
    }
}
